mkdocs/config/config_options.py

```diff
@@ -0,0 +1,333 @@
+import os
+
+import six
+
+from mkdocs import utils, legacy
+from mkdocs.legacy import OrderedDict
+
+
+class ValidationError(Exception):
+    """Raised during the validation process of the config on errors."""
+
+
+class BaseConfigOption(object):
+    """
+    The BaseConfigOption adds support for default values and required values
+
+    It then delegates the validation and (optional) post processing to
+    subclasses.
+    """
+
+    def __init__(self, default=None, required=False):
+        self.default = default
+        self.required = required
+        self.warnings = []
+
+    def is_required(self):
+        return self.required
+
+    def validate(self, value):
+        """
+        Perform some initial validation.
+
+        If the option is empty (None) and isn't required, leave it as such. If
+        it is empty but has a default, use that. Finally, call the _validate
+        method on the subclass unless.
+        """
+
+        if value is None:
+            if self.default is not None:
+                value = self.default
+            elif not self.required:
+                return
+            elif self.required:
+                raise ValidationError("Required configuration not provided.")
+
+        return self._validate(value)
+
+    def _validate(self, value):
+        """
+        Perform validation for a value.
+
+        The _validate method should be implemented by subclasses.
+        """
+        return value
+
+    def post_process(self, config, key_name):
+        """
+        After all options have passed validation, perform a post process to
```

Comment:
Yeah, that's why it gets the full config too. So it can do something based on another value. It's slightly limited, if you require the result of another `post_process` then you are outta luck :) but good enough for the cases I had.